snow/consensus/snowman/bootstrapper/majority.go

```diff
@@ -0,0 +1,196 @@
+)
+
+type majority struct {
+	log            logging.Logger
+	nodeWeights    map[ids.NodeID]uint64
+	maxOutstanding int
+
+	pendingSendAcceptedFrontier set.Set[ids.NodeID]
+	outstandingAcceptedFrontier set.Set[ids.NodeID]
+	receivedAcceptedFrontierSet set.Set[ids.ID]
+	receivedAcceptedFrontier    []ids.ID
+
+	pendingSendAccepted set.Set[ids.NodeID]
+	outstandingAccepted set.Set[ids.NodeID]
+	receivedAccepted    map[ids.ID]uint64
+	accepted            []ids.ID
+}
+
+func New(
+	log logging.Logger,
+	nodeWeights map[ids.NodeID]uint64,
+	maxFrontiers int,
+	maxOutstanding int,
+) (Bootstrapper, error) {
+	nodeIDs := maps.Keys(nodeWeights)
+	m := &majority{
+		log:                 log,
+		nodeWeights:         nodeWeights,
+		maxOutstanding:      maxOutstanding,
+		pendingSendAccepted: set.Of(nodeIDs...),
+		receivedAccepted:    make(map[ids.ID]uint64),
+	}
+
+	maxFrontiers = math.Min(maxFrontiers, len(nodeIDs))
+	sampler := sampler.NewUniform()
+	sampler.Initialize(uint64(len(nodeIDs)))
+	indicies, err := sampler.Sample(maxFrontiers)
+	for _, index := range indicies {
+		m.pendingSendAcceptedFrontier.Add(nodeIDs[index])
+	}
+
+	log.Debug("sampled nodes to seed bootstrapping frontier",
+		zap.Reflect("sampledNodes", m.pendingSendAcceptedFrontier),
+		zap.Int("numNodes", len(nodeIDs)),
+	)
+
+	return m, err
+}
+
+func (m *majority) GetAcceptedFrontiersToSend(context.Context) set.Set[ids.NodeID] {
+	return getPeersToSend(
+		&m.pendingSendAcceptedFrontier,
+		&m.outstandingAcceptedFrontier,
+		m.maxOutstanding,
+	)
+}
+
+func (m *majority) RecordAcceptedFrontier(_ context.Context, nodeID ids.NodeID, blkIDs ...ids.ID) {
+	if !m.outstandingAcceptedFrontier.Contains(nodeID) {
+		m.log.Error("received unexpected message",
+			zap.Stringer("messageOp", message.AcceptedFrontierOp),
+			zap.Stringer("nodeID", nodeID),
+			zap.Stringers("blkIDs", blkIDs),
+		)
+		return
+	}
+
+	m.outstandingAcceptedFrontier.Remove(nodeID)
+	m.receivedAcceptedFrontierSet.Add(blkIDs...)
+
+	if !m.finishedFetchingAcceptedFrontiers() {
+		return
+	}
+
+	m.receivedAcceptedFrontier = m.receivedAcceptedFrontierSet.List()
+
+	m.log.Debug("finalized bootstrapping frontier",
+		zap.Stringers("frontier", m.receivedAcceptedFrontier),
+	)
+}
+
+func (m *majority) GetAcceptedFrontier(context.Context) ([]ids.ID, bool) {
+	return m.receivedAcceptedFrontier, m.finishedFetchingAcceptedFrontiers()
+}
+
+func (m *majority) GetAcceptedToSend(context.Context) set.Set[ids.NodeID] {
+	if !m.finishedFetchingAcceptedFrontiers() {
+		return nil
+	}
+
+	return getPeersToSend(
+		&m.pendingSendAccepted,
+		&m.outstandingAccepted,
+		m.maxOutstanding,
+	)
+}
+
+func (m *majority) RecordAccepted(_ context.Context, nodeID ids.NodeID, blkIDs []ids.ID) error {
+	if !m.outstandingAccepted.Contains(nodeID) {
+		m.log.Error("received unexpected message",
+			zap.Stringer("messageOp", message.AcceptedOp),
+			zap.Stringer("nodeID", nodeID),
+			zap.Stringers("blkIDs", blkIDs),
+		)
```

Comment:
This used to be a DEBUG log because the engine used to need to handle unexpected messages... But these should be filtered by the chain router now.